nest-tax-backend/src/utils/subservices/line-logger.subservice.ts

```diff
@@ -0,0 +1,99 @@
+import * as process from 'node:process'
+
+import { LoggerService } from '@nestjs/common'
+
+import { escapeForLogfmt, isLogfmt, ToLogfmt } from '../logging'
+
+function getCurrentDateTime(): string {
+  const now = new Date()
+  return now.toLocaleString('en-US', {
+    year: 'numeric',
+    month: '2-digit',
+    day: '2-digit',
+    hour: '2-digit',
+    minute: '2-digit',
+    second: '2-digit',
+    hour12: true,
+  })
+}
+
+export class LineLoggerSubservice implements LoggerService {
+  protected context?: string
+
+  protected color: boolean
+
+  constructor(context?: string, color = true) {
+    this.context = context
+    this.color = color
+  }
+
+  private printLog(
+    severity: string,
+    message: unknown,
+    optionalParams: unknown[],
+    colorCode: string,
+  ) {
+    const completeArray = [message, ...optionalParams]
```

Comment:
While, creating an initially empty array is mostly OK, I would write it more in declarative style (you compose the code step by step and use pure functions and embraces immutability) like this:
```
private printLog(
  severity: string,
  message: unknown,
  optionalParams: unknown[],
  colorCode: string,
) {
  const completeArray = [message, ...optionalParams]

  const stringMessages = completeArray
    .filter((item): item is string => typeof item === 'string')
    .join(' ')

  const otherItems = completeArray
    .filter((item) => typeof item !== 'string')

  const formattedStringMessages = stringMessages.length > 0
    ? isLogfmt(stringMessages)
      ? ' '.concat(stringMessages)
      : ` message="${escapeForLogfmt(stringMessages)}"`
    : ''

  const formattedOtherItems = otherItems
    .map((param) => ToLogfmt(param))
    .join(' ')

  const formattedContext = this.context
    ? ` context="${this.context}"`
    : ''

  const colorStart = this.color ? colorCode : ''
  const colorEnd = this.color ? '\u001B[0m' : ''

  // eslint-disable-next-line no-console
  console.log(
    [
      colorStart,
      `process="[Nest]"`,
      `processPID="${process.pid}"`,
      `datetime="${getCurrentDateTime()}"`,
      `severity="${severity}"`,
      formattedContext,
      formattedStringMessages,
      formattedOtherItems,
      colorEnd,
    ]
      .filter(Boolean)
      .join(' ')
  )
}
```
You can see that:
1. This approach is TypeScript first. When creating variables initially you have to set type and then let the code match it. This will derive the types correctly from the code.
2. Everything is named and easier to follow, you create and name variable that explains what is happening. E.g. `stringItemsFormatted`, `otherItemsFormatted`.
3. Every step is pretty simple and does only one (transforms one kind of data to the another that leads to the result) as opposed to mutating, combining other items/string items logic
4. Console.log is built by joining the strings and more readable IMHO.

WDYT?